src/gpu/d3d11/SDL_gpu_d3d11.c

```diff
@@ -3106,11 +3118,52 @@ static void D3D11_GetTransferData(
 	}
 	else /* TEXTURE */
 	{
-		SDL_memcpy(
-			((Uint8*) data) + copyParams->dstOffset,
-			(Uint8*) buffer->textureTransfer.data + copyParams->srcOffset,
-			copyParams->size
-		);
+        SDL_LockMutex(renderer->contextLock);
+
+        /* Read from the staging texture */
+        res = ID3D11DeviceContext_Map(
+            renderer->immediateContext,
+            buffer->textureTransfer.downloadTexture,
+            0,
+            D3D11_MAP_READ,
+            0,
+            &subresource
+        );
+        ERROR_CHECK_RETURN("Could not map texture for reading",)
+
+        dataPtr = (Uint8*) data + copyParams->dstOffset;
+
+        if (buffer->textureTransfer.downloadTightlyPacked)
+        {
+            SDL_memcpy(
+                dataPtr,
+                (Uint8*) subresource.pData + copyParams->srcOffset,
+                buffer->textureTransfer.downloadDepth * buffer->textureTransfer.downloadHeight * buffer->textureTransfer.downloadBytesPerRow
+            );
+        }
+        else
+        {
+            for (Uint32 depth = 0; depth < buffer->textureTransfer.downloadDepth; depth += 1)
+            {
+                for (Uint32 row = 0; row < buffer->textureTransfer.downloadHeight; row += 1)
+                {
+                    SDL_memcpy(
+                        dataPtr,
+                        (Uint8*) subresource.pData + copyParams->srcOffset + (depth * buffer->textureTransfer.downloadBytesPerDepthSlice) + (row * buffer->textureTransfer.downloadBytesPerRow),
+                        buffer->textureTransfer.downloadBytesPerRow
+                    );
+                    dataPtr += buffer->textureTransfer.downloadBytesPerRow;
```

Comment:
And if we copy less than downloadBytesPerRow (w * bpp i guess?) we would probably want to advance the destination pointer less too.

Reply:
downloadBytesPerRow is the specific stride value the client has requested. I hadn't really considered that copyParams->size isn't being respected here though. 

Probably worth considering whether supporting stride in readback is even worth anything.